api/auth/azure/azure.go

```diff
@@ -0,0 +1,204 @@
+package azure
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"net/url"
+
+	"github.com/hashicorp/vault/api"
+)
+
+type AzureAuth struct {
+	roleName  string
+	mountPath string
+	resource  string
+}
+
+var _ api.AuthMethod = (*AzureAuth)(nil)
+
+type LoginOption func(a *AzureAuth) error
+
+type responseJSON struct {
+	AccessToken  string `json:"access_token"`
+	RefreshToken string `json:"refresh_token"`
+	ExpiresIn    string `json:"expires_in"`
+	ExpiresOn    string `json:"expires_on"`
+	NotBefore    string `json:"not_before"`
+	Resource     string `json:"resource"`
+	TokenType    string `json:"token_type"`
+}
+
+type metadataJSON struct {
+	Compute computeJSON `json:"compute"`
+}
+
+type computeJSON struct {
+	VMName            string `json:"name"`
+	VMScaleSetName    string `json:"vmScaleSetName"`
+	SubscriptionID    string `json:"subscriptionId"`
+	ResourceGroupName string `json:"resourceGroupName"`
+}
+
+const (
+	defaultMountPath     = "azure"
+	defaultResourceURL   = "https://management.azure.com"
+	metadataEndpoint     = "http://169.254.169.254"
+	metadataAPIVersion   = "2018-02-01"
```

Comment:
nit: It should be safe to use a newer api version to avoid any issues if Microsoft starts deprecating & removing older versions. List of versions can be found [here](https://docs.microsoft.com/en-us/azure/virtual-machines/linux/instance-metadata-service?tabs=linux#supported-api-versions).

Reply:
Good point, thank you!